### PR TITLE
Support explicit TRT profiles from provider options

### PR DESCRIPTION
**Pull request by chilo-ms**
Previous behavior of TRT EP to set TRT optimization profiles for dynamic shape input is based on input tensor values. Users can't explicitly specify the profiles.

This PR makes users capable of specifying min/max/opt profiles through newly added three provider options:

`trt_profile_min_shapes`, `trt_profile_max_shapes` and `trt_profile_opt_shapes` 
with the format of "input1:dim1xdim2...,input2:dim3xdim4...".
(Note: It's similar to --minShapes, --maxShapes and --optShapes of trtexec command-line [flags](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#trtexec-flags))

For example, if you are using onnxruntime_perf_test, you can try this:

`./onnxruntime_perf_test -e tensorrt -r 1 -i "trt_profile_min_shapes|imgs:1x3x384x288 trt_profile_max_shapes|imgs:32x3x384x288 trt_profile_opt_shapes|imgs:16x3x384x288" your_model_path`

If the engine cache is enabled, you still need to provide these three explicit provider options in order to use this feature. ORT TRT will compare the min/max/opt profile shape with the ones saved in .profile file to decide whether to rebuild the engine.

Constraints to use these provider options: (1) Need to specify min/max/opt profile shapes for all the dynamic shape input 

 

This feature is also requested by other users:
https://github.com/microsoft/onnxruntime/issues/13851

### CODE REVIEW

**Comment by jywu-msft** [comment]
needs to resolve against conflicts?

**Comment by tianleiwu** [comment]
What if user specify min shape but not max shape for an input? We can force full input profile by designing interface as ORT_TENSORRT_INPUT_PROFILE: "input1_name|min_shape|opt_shape|max_shape,input2_name|min_shape|opt_shape|max_shape,...".

**Comment by chilo-ms** [comment]
> We can force full input profile by designing interface as ORT_TENSORRT_INPUT_PROFILE: "input1_name:min_shape|opt_shape|max_shape,input2_name:min_shape|opt_shape|max_shape,...".

That could be a way to design the interface. But we want to make the provider option consistent with trtexec where it uses --minShapes, --maxShapes and --optShapes as described in this PR description.

File: onnxruntime/test/perftest/ort_test_session.cc [hunk 1]
```diff
@@ -341,8 +345,34 @@ OnnxRuntimeTestSession::OnnxRuntimeTestSession(Ort::Env& env, std::random_device
...
+        if (!value.empty()) {
+          trt_profile_max_shapes = value;
+        } else {
+          ORT_THROW("[ERROR] [TensorRT] The value for the key 'trt_profile_max_shapes' should be a non-emtpy string.\n");
+        }
+      } else if (key == "trt_profile_opt_shapes") {
+        if (!value.empty()) {
+          trt_profile_opt_shapes = value;
+        } else {
+          ORT_THROW("[ERROR] [TensorRT] The value for the key 'trt_profile_opt_shapes' should be a non-emtpy string.\n");
```
Comment (tianleiwu) [inline onnxruntime/test/perftest/ort_test_session.cc:364]: typo: non-emtpy => non-empty.
Please search and fix other places as well.

**Comment by tianleiwu** [comment]
> That could be another way to design the interface. But we want to make the provider options be consistent with trtexec where it uses --minShapes, --maxShapes and --optShapes as described in this PR description.

That sounds reasonable. It is fine to keep current design.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 2]
```diff
@@ -554,6 +804,15 @@ TensorrtExecutionProvider::TensorrtExecutionProvider(const TensorrtExecutionProv
...
       tactic_sources_ = tactic_sources_env;
     }
+
+    profile_min_shapes = onnxruntime::GetEnvironmentVar(tensorrt_env_vars::kProfilesMinShapes);
+    profile_max_shapes = onnxruntime::GetEnvironmentVar(tensorrt_env_vars::kProfilesMaxShapes);
+    profile_opt_shapes = onnxruntime::GetEnvironmentVar(tensorrt_env_vars::kProfilesOptShapes);
+
+    const std::string engine_cache_built_with_explicit_profiles_env = onnxruntime::GetEnvironmentVar(tensorrt_env_vars::kEngineCacheBuiltWithExplicitProfiles);
+    if (!engine_cache_built_with_explicit_profiles_env.empty()) {
+      engine_cache_built_with_explicit_profiles_ = (std::stoi(engine_cache_built_with_explicit_profiles_env) == 0 ? false : true);
```
[the remark this answers is not in the document]
Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: If we change `engine_cache_built_with_explicit_profiles_env  == "0" ? false : true`, if user accidentally set non-integer string to this environment variable, TRT EP will interpret as true. That seems not right. So, I added try/catch to stoi() handle the exception.

File: include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h [hunk 3]
```diff
@@ -41,7 +41,8 @@ struct OrtTensorRTProviderOptionsV2 {
...
                                                 // tactic sources (default = all available tactics) e.g. "-CUDNN,+CUBLAS" available keys: "CUBLAS"|"CUBLAS_LT"|"CUDNN"|"EDGE_MASK_CONVOLUTIONS"
   const char* trt_extra_plugin_lib_paths;       // specify extra TensorRT plugin library paths
-  const char* trt_profile_min_shapes;          // Specify the range of the input shapes to build the engine with
-  const char* trt_profile_max_shapes;          // Specify the range of the input shapes to build the engine with
-  const char* trt_profile_opt_shapes;          // Specify the range of the input shapes to build the engine with
+  const char* trt_profile_min_shapes;           // Specify the range of the input shapes to build the engine with
+  const char* trt_profile_max_shapes;           // Specify the range of the input shapes to build the engine with
+  const char* trt_profile_opt_shapes;           // Specify the range of the input shapes to build the engine with
+  int trt_engine_cache_built_with_explicit_profiles; // Whether the engine cache is built with explicit profiles provided by user. Default 0 = false, nonzero = true
```
[the remark this answers is not in the document]
Comment (chilo-ms) [reply include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:47]: Re: I think the engine shall store the profile that used to build it. If profile changes, we shall rebuild the engine.

I think this is a good suggestion. Previously, TRT EP depends on a .profile cache file (generated by TRT EP when building engine cache) to see the min/max/opt shape ranges and decide whether to update the range (according to the input tensor) as well as rebuild the engine if needed. Due to this old design, if we don't have this flag, TRT EP doesn't know the engine is built with explicit profiles and will go the old code path and rebuild the engine.

Let me think about how to leverage TRT engine APIs to get the profiles info. Thanks.

[the remark this answers is not in the document]
Comment (tianleiwu) [reply include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:47]: If user provides explicit profiles, we still could still compare with .profile cache file to see whether profile has changed. If there is no profile cache file or profile has changes, we can rebuild the engine and update .profile cache file.

[the remark this answers is not in the document]
Comment (chilo-ms) [reply include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:47]: Let's consider the case where engine cache doesn't need to rebuild. For example, if user has engine cache (built with explicit profiles on one machine) and want to deploy the engine to another machine. 
In this case, I prefer user doesn't need to explicitly specify the profiles again, so we don't know whether the profiles is changed or not. We can force users to explicitly set profiles, but I think it's a little bit complex to use.

**Comment by jywu-msft** [comment]
lint errors https://github.com/microsoft/onnxruntime/actions/runs/4837212518/jobs/8620872293?pr=15546

**Comment by chilo-ms** [comment]
> lint errors https://github.com/microsoft/onnxruntime/actions/runs/4837212518/jobs/8620872293?pr=15546

Yes, will fix the format warnings

**Review by jywu-msft — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**